// @alfresco/adf-core/esm5/pipes/decimal-number.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { DecimalPipe } from '@angular/common';
import { Pipe } from '@angular/core';
import { AppConfigService } from '../app-config/app-config.service';
import { UserPreferencesService, UserPreferenceValues } from '../services/user-preferences.service';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
var DecimalNumberPipe = /** @class */ (function () {
    function DecimalNumberPipe(userPreferenceService, appConfig) {
        var _this = this;
        this.userPreferenceService = userPreferenceService;
        this.appConfig = appConfig;
        this.defaultLocale = DecimalNumberPipe.DEFAULT_LOCALE;
        this.defaultMinIntegerDigits = DecimalNumberPipe.DEFAULT_MIN_INTEGER_DIGITS;
        this.defaultMinFractionDigits = DecimalNumberPipe.DEFAULT_MIN_FRACTION_DIGITS;
        this.defaultMaxFractionDigits = DecimalNumberPipe.DEFAULT_MAX_FRACTION_DIGITS;
        this.onDestroy$ = new Subject();
        if (this.userPreferenceService) {
            this.userPreferenceService.select(UserPreferenceValues.Locale)
                .pipe(takeUntil(this.onDestroy$))
                .subscribe((/**
             * @param {?} locale
             * @return {?}
             */
            function (locale) {
                if (locale) {
                    _this.defaultLocale = locale;
                }
            }));
        }
        if (this.appConfig) {
            this.defaultMinIntegerDigits = this.appConfig.get('decimalValues.minIntegerDigits', DecimalNumberPipe.DEFAULT_MIN_INTEGER_DIGITS);
            this.defaultMinFractionDigits = this.appConfig.get('decimalValues.minFractionDigits', DecimalNumberPipe.DEFAULT_MIN_FRACTION_DIGITS);
            this.defaultMaxFractionDigits = this.appConfig.get('decimalValues.maxFractionDigits', DecimalNumberPipe.DEFAULT_MAX_FRACTION_DIGITS);
        }
    }
    /**
     * @param {?} value
     * @param {?=} digitsInfo
     * @param {?=} locale
     * @return {?}
     */
    DecimalNumberPipe.prototype.transform = /**
     * @param {?} value
     * @param {?=} digitsInfo
     * @param {?=} locale
     * @return {?}
     */
    function (value, digitsInfo, locale) {
        /** @type {?} */
        var actualMinIntegerDigits = digitsInfo && digitsInfo.minIntegerDigits ? digitsInfo.minIntegerDigits : this.defaultMinIntegerDigits;
        /** @type {?} */
        var actualMinFractionDigits = digitsInfo && digitsInfo.minFractionDigits ? digitsInfo.minFractionDigits : this.defaultMinFractionDigits;
        /** @type {?} */
        var actualMaxFractionDigits = digitsInfo && digitsInfo.maxFractionDigits ? digitsInfo.maxFractionDigits : this.defaultMaxFractionDigits;
        /** @type {?} */
        var actualDigitsInfo = actualMinIntegerDigits + "." + actualMinFractionDigits + "-" + actualMaxFractionDigits;
        /** @type {?} */
        var actualLocale = locale || this.defaultLocale;
        /** @type {?} */
        var datePipe = new DecimalPipe(actualLocale);
        return datePipe.transform(value, actualDigitsInfo);
    };
    /**
     * @return {?}
     */
    DecimalNumberPipe.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.onDestroy$.next(true);
        this.onDestroy$.complete();
    };
    DecimalNumberPipe.DEFAULT_LOCALE = 'en-US';
    DecimalNumberPipe.DEFAULT_MIN_INTEGER_DIGITS = 1;
    DecimalNumberPipe.DEFAULT_MIN_FRACTION_DIGITS = 0;
    DecimalNumberPipe.DEFAULT_MAX_FRACTION_DIGITS = 2;
    DecimalNumberPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'adfDecimalNumber',
                    pure: false
                },] }
    ];
    /** @nocollapse */
    DecimalNumberPipe.ctorParameters = function () { return [
        { type: UserPreferencesService },
        { type: AppConfigService }
    ]; };
    return DecimalNumberPipe;
}());
export { DecimalNumberPipe };
if (false) {
    /** @type {?} */
    DecimalNumberPipe.DEFAULT_LOCALE;
    /** @type {?} */
    DecimalNumberPipe.DEFAULT_MIN_INTEGER_DIGITS;
    /** @type {?} */
    DecimalNumberPipe.DEFAULT_MIN_FRACTION_DIGITS;
    /** @type {?} */
    DecimalNumberPipe.DEFAULT_MAX_FRACTION_DIGITS;
    /** @type {?} */
    DecimalNumberPipe.prototype.defaultLocale;
    /** @type {?} */
    DecimalNumberPipe.prototype.defaultMinIntegerDigits;
    /** @type {?} */
    DecimalNumberPipe.prototype.defaultMinFractionDigits;
    /** @type {?} */
    DecimalNumberPipe.prototype.defaultMaxFractionDigits;
    /** @type {?} */
    DecimalNumberPipe.prototype.onDestroy$;
    /** @type {?} */
    DecimalNumberPipe.prototype.userPreferenceService;
    /** @type {?} */
    DecimalNumberPipe.prototype.appConfig;
}
