// @alfresco/adf-core/esm5/userinfo/models/identity-group.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var /*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
IdentityGroupModel = /** @class */ (function () {
    function IdentityGroupModel(obj) {
        if (obj) {
            this.id = obj.id || null;
            this.name = obj.name || null;
            this.path = obj.path || null;
            this.realmRoles = obj.realmRoles || null;
            this.clientRoles = obj.clientRoles || null;
            this.access = obj.access || null;
            this.attributes = obj.attributes || null;
        }
    }
    return IdentityGroupModel;
}());
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
export { IdentityGroupModel };
if (false) {
    /** @type {?} */
    IdentityGroupModel.prototype.id;
    /** @type {?} */
    IdentityGroupModel.prototype.name;
    /** @type {?} */
    IdentityGroupModel.prototype.path;
    /** @type {?} */
    IdentityGroupModel.prototype.realmRoles;
    /** @type {?} */
    IdentityGroupModel.prototype.clientRoles;
    /** @type {?} */
    IdentityGroupModel.prototype.access;
    /** @type {?} */
    IdentityGroupModel.prototype.attributes;
}
