// @alfresco/adf-content-services/esm5/content-node-selector/name-location-cell/name-location-cell.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Input, ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
var NameLocationCellComponent = /** @class */ (function () {
    function NameLocationCellComponent() {
        this.name = '';
        this.path = '';
    }
    /**
     * @return {?}
     */
    NameLocationCellComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.row) {
            this.name = this.row.getValue('name');
            /** @type {?} */
            var fullPath = this.row.getValue('path');
            if (fullPath) {
                this.path = fullPath.name || '';
            }
        }
    };
    NameLocationCellComponent.decorators = [
        { type: Component, args: [{
                    selector: 'adf-name-location-cell',
                    template: "\n        <div class=\"adf-name-location-cell-name adf-datatable-cell-value\" [title]=\"name\">{{ name }}</div>\n        <div class=\"adf-name-location-cell-location adf-datatable-cell-value\" [title]=\"path\">{{ path }}</div>\n    ",
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: { class: 'adf-name-location-cell adf-datatable-content-cell' },
                    styles: [""]
                }] }
    ];
    NameLocationCellComponent.propDecorators = {
        row: [{ type: Input }]
    };
    return NameLocationCellComponent;
}());
export { NameLocationCellComponent };
if (false) {
    /** @type {?} */
    NameLocationCellComponent.prototype.name;
    /** @type {?} */
    NameLocationCellComponent.prototype.path;
    /** @type {?} */
    NameLocationCellComponent.prototype.row;
}
