// @alfresco/adf-core/esm2015/userinfo/models/identity-user.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2019 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
export class IdentityUserModel {
    /**
     * @param {?=} obj
     */
    constructor(obj) {
        if (obj) {
            this.id = obj.id || null;
            this.firstName = obj.firstName || null;
            this.lastName = obj.lastName || null;
            this.email = obj.email || null;
            this.username = obj.username || null;
            this.createdTimestamp = obj.createdTimestamp || null;
            this.emailVerified = obj.emailVerified || null;
            this.enabled = obj.enabled || null;
        }
    }
}
if (false) {
    /** @type {?} */
    IdentityUserModel.prototype.id;
    /** @type {?} */
    IdentityUserModel.prototype.firstName;
    /** @type {?} */
    IdentityUserModel.prototype.lastName;
    /** @type {?} */
    IdentityUserModel.prototype.email;
    /** @type {?} */
    IdentityUserModel.prototype.username;
    /** @type {?} */
    IdentityUserModel.prototype.createdTimestamp;
    /** @type {?} */
    IdentityUserModel.prototype.emailVerified;
    /** @type {?} */
    IdentityUserModel.prototype.enabled;
}
export class IdentityUserPasswordModel {
    /**
     * @param {?=} obj
     */
    constructor(obj) {
        if (obj) {
            this.type = obj.type;
            this.value = obj.value;
            this.temporary = obj.temporary;
        }
    }
}
if (false) {
    /** @type {?} */
    IdentityUserPasswordModel.prototype.type;
    /** @type {?} */
    IdentityUserPasswordModel.prototype.value;
    /** @type {?} */
    IdentityUserPasswordModel.prototype.temporary;
}
/**
 * @record
 */
export function IdentityUserQueryResponse() { }
if (false) {
    /** @type {?} */
    IdentityUserQueryResponse.prototype.entries;
    /** @type {?} */
    IdentityUserQueryResponse.prototype.pagination;
}
export class IdentityUserQueryCloudRequestModel {
    /**
     * @param {?=} obj
     */
    constructor(obj) {
        if (obj) {
            this.first = obj.first;
            this.max = obj.max;
        }
    }
}
if (false) {
    /** @type {?} */
    IdentityUserQueryCloudRequestModel.prototype.first;
    /** @type {?} */
    IdentityUserQueryCloudRequestModel.prototype.max;
}
export class IdentityJoinGroupRequestModel {
    /**
     * @param {?=} obj
     */
    constructor(obj) {
        if (obj) {
            this.realm = obj.realm;
            this.userId = obj.userId;
            this.groupId = obj.groupId;
        }
    }
}
if (false) {
    /** @type {?} */
    IdentityJoinGroupRequestModel.prototype.realm;
    /** @type {?} */
    IdentityJoinGroupRequestModel.prototype.userId;
    /** @type {?} */
    IdentityJoinGroupRequestModel.prototype.groupId;
}
